feat: add RecordRunModal component

astone123 opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes #23764

### User facing changelog
<!-- 
Explain the change(s) for every user to read in our changelog. Examples: https://on.cypress.io/changelog
If the change is not user-facing, write "n/a".
-->

n/a

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

This PR adds the `RecordRunModal` UI component and refactors some logic from `RecordBanner` to be reusable between these two components. The work to integrate this component into the UI will be done in #23766

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

The component isn't actually used anywhere in the UI yet, so take a look at the test

1. Run `RecordRunModal.cy.tsx` in `frontend-shared`
2. Look at the component and ensure that it meets all of the criteria laid out in the issue
3. Verify that `RecordBanner` still works as before

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

n/a

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

astone123 committed:
feat: add RecordRunModal component

astone123 committed:
feat: add missing assertion

astone123 committed:
feat: import using @packages syntax

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordPrompt.vue:
```diff
@@ -0,0 +1,53 @@
+const { t } = useI18n()
+
+gql`
+query RecordPrompt {
```
I *think* it makes sense for modals to have their own `useQuery` (as opposed to fragment + props).

I read https://relay.dev/docs/principles-and-architecture/thinking-in-relay/ in more delay and it seems FB, at least, are heavily using the "big prop object + single query" approach... I see the value in this, but also interested in exploring the "`useQuery` + fetch from the cache" too, and seeing what works best for our use case.

mike-plummer replied:
I'm a bit torn on it in this case. I'm in favor of using a separate `useQuery` when we aren't 100% sure we'll need the data or if the data is expensive (e.g, cloud project runs) or may need to be re-fetched, but in this case the data is technically cloud data but it's pretty lightweight and other parts of the app already fetch it. Seems like encapsulating the query here really just makes this difficult to test.

Personally I would wrap this with a data-managing component with the sole responsibility of doing the query and mapping the result to props, then this component would just accept props and be very simple to test the actual input => output behaviors of without wrangling GQL mocks

lmiller1990 replied:
>  I'm in favor of using a separate useQuery when we aren't 100% sure we'll need the data or if the data is expensive (e.g, cloud project runs) or may need to be re-fetched

This was what I was thinking. I can see an argument for "keep doing it how we do it", though. I'm not really sold on changing it **just** for the purpose of testing - we could easily just use `cy.intercept` to test the modal.

Could go either way - I've seen code bases making use of `useQuery` everywhere (mainly Apollo based one) and others doing the "one big prop" pattern (mainly Relay). I guess it's just competing philosophies... in which case, I wonder if we should go with "one big prop" consistently?

Definitely good to standardize how we'd like to handle this, document it, and even considering linting for it.

lmiller1990 replied:
@astone123 thoughts? Maybe we do want to stick with the single query/prop pattern for now?

astone123 replied:
I ended up separating the gql logic out into an "adapter" component and then passing the data down into `RecordPrompt`. I also wrote separate component tests for `RecordPrompt`

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordPrompt.vue:
```diff
@@ -0,0 +1,53 @@
+})
+
+const recordCommand = computed(() => {
+  const componentFlagOrSpace = query.data?.value?.currentProject?.currentTestingType === 'component' ? ' --component ' : ' '
```
```suggestion
  const componentFlagOrSpace = query.data.value?.currentProject?.currentTestingType === 'component' ? ' --component ' : ' '
```

Nit... I think `data` is always defined and doesn't need `?.`.

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordRunModal.cy.tsx:
```diff
@@ -0,0 +1,34 @@
+    cy.percySnapshot()
+  })
+
+  it('calls close when X is clicked', () => {
```
Not sure if it's covered somewhere else but do we have coverage for "closes standard modal when ESC is pressed"? I see that as a requirement, not sure if we have the coverage -- if we don''t might be a good time to add it.

astone123 replied:
We do have coverage for this in `StandardModal.cy.tsx`
```
    it('closes with Esc key', () => {
      cy.get('body').type('{esc}')
      cy.get('@updateSpy').should('have.been.calledOnceWith', false)
    })
```

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordRunModal.vue:
```diff
@@ -0,0 +1,36 @@
+    class="transition transition-all duration-200"
+    variant="bare"
+    :title="t('specPage.banners.record.title')"
+    :model-value="isModalOpen"
+    :help-link="helpLink"
+    :no-help="!helpLink"
+    data-cy="record-run-modal"
+    @update:model-value="close"
```
I don't think we are doing `v-model` correctly (anywhere, lol, we should make this change everywhere). [Docs on v-model](https://vuejs.org/guide/components/events.html#usage-with-v-model).

```suggestion
    :help-link="helpLink"
    :no-help="!helpLink"
    data-cy="record-run-modal"
    v-model="isModalOpen"
```

`v-model` is basically a shorthand for:

```html
<CustomInput
  :modelValue="searchText"
  @update:modelValue="newValue => searchText = newValue"
/>

<!-- can be simplifed -->
<CustomInput v-model="searchText" />
```

astone123 replied:
When I try this I get an ESLint error where I try to pass `v-model="isModalOpen"`
```
Unexpected mutation of "isModalOpen" prop.
```
and my test fails because the `close` prop function is never called when I click X 🤔

lmiller1990 replied:
Hm ok... weird. I think `<StandardModal>` needs a bit of a refactor. I think

```html
<StandardModal 
  :open="open"
  @close="close"
/>
```

Seems more readable than `@update:model-value="close"`. Either way, happy to go with this for now, seems this is the pattern we are using everywhere else.

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordRunModal.vue:
```diff
@@ -0,0 +1,36 @@
+}>()
+
+const helpLink =
+`https://on.cypress.io/cypress-run-record-key?utm_medium="${props.utmMedium}",utm_source="${getUtmSource()}"${props.utmContent ? `,utm_content="${props.utmContent}"` : ''}`
```
Will any of these props values ever change? If so, we'd want this to `computed`.

astone123 replied:
I'm not sure what the answer is yet... I think we can handle that during the implementation

lmiller1990 commented on packages/app/src/specs/banners/RecordBanner.vue:
```diff
@@ -16,47 +16,16 @@
-  }
-}
-`
+import RecordPrompt from '@packages/frontend-shared/src/gql-components/RecordPrompt.vue'
```
Should this be in `packages/app`? I don't think we will ever use the `RecordPrompt` in Launchpad.

astone123 replied:
I put it in `frontend-shared` because that's where `Auth` is, which I think is where `RecordRunModal` will be used 🤔 not sure where to put it

mike-plummer replied:
If I understand the linked issue correctly the intent is to eventually give users a push to record as part of the login flow (login modal => connect modal => record modal), so it likely will need to be a shared component since the login modal is used in launchpad & app

lmiller1990 commented on packages/frontend-shared/src/gql-components/RecordPrompt.vue:
```diff
@@ -0,0 +1,53 @@
+  return (query.data?.value?.currentProject?.cloudProject?.__typename === 'CloudProject' && query.data?.value?.currentProject.cloudProject.recordKeys?.[0]?.key) ?? '<record-key>'
+})
+
+const recordCommand = computed(() => {
```
Seems we are missing coverage around this, or are we happy to rely on Percy for testing this?

astone123 replied:
We cover that in `RecordBanner.cy.tsx`, so I think just covering it with Percy here is fine

lmiller1990 reviewed (commented):
Some minor comments here and there :+1:

astone123 committed:
feat: remove unnecessary null check

mike-plummer commented on packages/frontend-shared/src/gql-components/RecordRunModal.vue:
```diff
@@ -0,0 +1,36 @@
+}>()
+
+const helpLink =
+`https://on.cypress.io/cypress-run-record-key?utm_medium="${props.utmMedium}",utm_source="${getUtmSource()}"${props.utmContent ? `,utm_content="${props.utmContent}"` : ''}`
```
Are we sure about the formatting of this? Should the query params really be separated by `,` instead of `&`? Why not use the existing `getUrlWithParams` utility function to build this?

astone123 replied:
Oh wow, yeah having `,` as the separator was a mistake. I didn't see that there was a utility function, will definitely use that 👍🏻

astone123 committed:
feat: move record key gql logic to separate component

astone123 committed:
feat: add another test case

astone123 committed:
Merge branch 'develop' into astone123/aci-record-run-modal-ui

warrensplayer commented on packages/app/src/specs/banners/RecordBanner.vue:
```diff
@@ -16,47 +16,16 @@
     }"
     @update:model-value="value => emit('update:modelValue', value)"
   >
-    <p class="mb-24px">
```
This copy should remain part of the banner. The RecordPrompt should just encapsulate the logic for creating the prompt itself.  That way, different places in the app that use it can have different copy surrounding the prompt.

warrensplayer commented on packages/frontend-shared/src/gql-components/RecordPrompt.vue:
```diff
@@ -0,0 +1,33 @@
+<template>
+  <p class="mb-24px">
```
Leave the copy block out of the RecordPrompt component so that the surrounding copy can be customized where it is used.

warrensplayer commented on packages/frontend-shared/src/gql-components/RecordRunModal.vue:
```diff
@@ -0,0 +1,43 @@
+    data-cy="record-run-modal"
+    @update:model-value="close"
+  >
+    <div class="max-w-175 py-7 px-6 text-gray-600">
```
Add a copy block here that would have the same content as the banner for this implementation.  This allows the `RecordPrompt` to just encapsulate the logic around the TerminalPrompt and the logic to create it for recording a run.

warrensplayer reviewed (changes requested):
The `RunsEmpty.vue` component has the same logic to construct a TerminalPrompt for recording runs.  It should also be refactored to use the new RecordPrompt component.  To accomplish that, it would require a change to the RecordPrompt component to pull out the copy part above the prompt.  I have noted that in other comments within the code.

astone123 committed:
feat: refactor RunsEmpty component

astone123 committed:
Merge branch 'astone123/aci-record-run-modal-ui' of github.com:cypress-io/cypress into astone123/aci-record-run-modal-ui

astone123 committed:
feat: update test

astone123 committed:
Merge branch 'develop' into astone123/aci-record-run-modal-ui

warrensplayer reviewed (changes requested):
The `RunsContainer.vue` is still spreading the `RunsEmpty` fragment on line 82 that no longer exists.

mike-plummer commented on packages/app/src/runs/RunsEmpty.cy.tsx:
```diff
@@ -1,14 +1,25 @@
+      } as any,
+    } as any
+
+    cy.contains('Record your first run').should('be.visible')
```
nitpick: I think elsewhere we're trying to avoid hardcoding strings and instead use the message bundle:

```
import defaultMessages from '@packages/frontend-shared/src/locales/en-US.json'
...
cy.contains(defaultMessages.specPage.banners.record.title).should('be.visible')
```

lmiller1990 replied:
It looks like a mixed bag - some tests hard-code it.

I find the hard-coded strings version of the test *much* more readable and easier to search for. I guess we should probably standardize this, though.

astone123 replied:
@lmiller1990 same, I wonder if there's a way that we could use the i18n Ally plugin to show the actual messages in the test code like it does for Vue components 🤔

astone123 committed:
feat: remove unused query

astone123 committed:
:wq
Merge branch 'astone123/aci-record-run-modal-ui' of github.com:cypress-io/cypress into astone123/aci-record-run-modal-ui

mike-plummer commented on packages/frontend-shared/src/gql-components/RecordRunModal.cy.tsx:
```diff
@@ -0,0 +1,40 @@
+    cy.mount(<RecordRunModalVue isModalOpen={true} close={cy.stub()} utmMedium="Nav" />)
+
+    cy.contains('Need help').should('have.attr', 'href', 'https://on.cypress.io/cypress-run-record-key?utm_medium=Nav&utm_source=Binary%3A+Launchpad&utm_content=')
+  })
```
Can we add a test case for if the gql query fails? Seems like we may end up with an empty/invalid record command if we fail to retrieve record keys for whatever reason

astone123 replied:
I do have a test in `RecordPrompt.cy.tsx` that I think covers this case. If we fail to retrieve a record key then we just put `<record-key>` in the command
```
  it('renders with no props', () => {
    cy.mount(<RecordPromptVue />)

    cy.findByText('npx cypress run --record --key <record-key>').should('be.visible')
  })
```

lmiller1990 reviewed (approved):
Seems good to me, just need to address a few nits. I'll approve, assuming those are going to get addressed :heavy_check_mark:

astone123 committed:
feat: use translation strings in tests

astone123 committed:
feat: assert that command shows in RecordRunModal

astone123 committed:
Merge branch 'develop' into astone123/aci-record-run-modal-ui

astone123 committed:
Merge branch 'develop' into astone123/aci-record-run-modal-ui

lmiller1990 committed:
Merge branch 'develop' into astone123/aci-record-run-modal-ui
